Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs

```diff
@@ -0,0 +1,142 @@
+            }
+
+            var deserializer = new PipGraphFragmentSerializer();
+            Task<bool> readFragmentTask = Task.Run(() =>
+            {
+                Task.WaitAll(dependencyNames.Select(dependencyName => m_readFragmentTasks[dependencyName].Item2).ToArray());
+                return deserializer.Deserialize(fragmentName, m_context, m_fragmentContext, filePath, (Pip p) => AddPipToGraph(fragmentName, p));
+            });
+
+            m_readFragmentTasks[fragmentName] = (deserializer, readFragmentTask);
+            return readFragmentTask;
+        }
+
+        /// <summary>
+        /// GetAllFragmentTasks
+        /// </summary>
+        public IReadOnlyCollection<(PipGraphFragmentSerializer, Task<bool>)> GetAllFragmentTasks()
+        {
+            return m_readFragmentTasks.Select(x => x.Value).ToList();
+        }
+
+        private bool AddPipToGraph(string fragmentName, Pip pip)
+        {
+            try
+            {
+                PipId originalPipId = pip.PipId;
+                pip.ResetPipIdForTesting();
+                bool added = false;
+                switch (pip.PipType)
+                {
+                    case PipType.Module:
+                        var modulePip = pip as ModulePip;
+                        added = m_pipGraph.AddModule(modulePip);
+                        break;
+                    case PipType.SpecFile:
+                        var specFilePip = pip as SpecFilePip;
+                        added = m_pipGraph.AddSpecFile(specFilePip);
+                        break;
+                    case PipType.Value:
+                        var valuePIp = pip as ValuePip;
+                        added = m_pipGraph.AddOutputValue(valuePIp);
+                        break;
+                    case PipType.Process:
+                        var p = pip as Process;
+                        added = m_pipGraph.AddProcess(p, default);
+                        if (p.IsService)
+                        {
+                            m_fragmentContext.AddPipIdMapping(originalPipId.Value, p.PipId.Value);
+                        }
+
+                        break;
+                    case PipType.CopyFile:
+                        var copyFile = pip as CopyFile;
+                        added = m_pipGraph.AddCopyFile(copyFile, default);
+                        break;
+                    case PipType.WriteFile:
+                        var writeFile = pip as WriteFile;
+                        added = m_pipGraph.AddWriteFile(writeFile, default);
+                        break;
+                    case PipType.SealDirectory:
+                        var sealDirectory = pip as SealDirectory;
+                        if (sealDirectory.Kind == SealDirectoryKind.Opaque || sealDirectory.Kind == SealDirectoryKind.SharedOpaque)
+                        {
+                            return true;

```

Comment:
why is it ok to skip over opaque directories?

Reply:
PipBuilder constructs those itself from the output directories listed in addprocess

---
In reply to: [289548569](https://github.com/microsoft/BuildXL/pull/394#discussion_r289548569) [](ancestors = 289548569)